### PR TITLE
NPM documentation

### PR DESCRIPTION
**Pull request by mosabua**
Ported and adapted from the Nexus 2 docs. This can also be used as a test guide.

https://issues.sonatype.org/browse/NEXUS-9220

Rendered output as the site artifact of the branch bamboo builds http://bamboo.s/browse/BOOK-NEXUS38

### CODE REVIEW

File: chapter-npm.asciidoc [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+The command line tool +npm+ is a package management solution for Javascript-based development. It is used to create and
+use 'node packaged modules' and is built into the popular Javascript platform http://www.nodejs.org/[Node.js], which is
+mostly used for server-side application development.
+
+The npm registry at https://www.npmjs.org/[https://www.npmjs.org/] is the default package registry, from which
+components can be retrieved.  It contains a large number of open source packages for Node.js based server-side
+application development, build tools like 'bower' or 'grunt' and many other packages for a variety of use cases.
+
+Nexus supports the npm registry format for proxy repositories. This allows you to take advantage of the packages in the
+npm registry and other public registries without incurring repeated downloads of packages, since the will be proxied in
```
Comment (joedragons) [inline chapter-npm.asciidoc]: "the will" -> they will

File: chapter-npm.asciidoc [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+Nexus.
+
+In addition, Nexus supports running your own private registry - also known as a hosted repository using the +npm+
```
Comment (joedragons) [inline chapter-npm.asciidoc]: I know comma = drama but suggest "registry, also known as a hosted repository, using".  The dash to me infers a seperate thought but really it's an interruption.  Am not keen on the double dash though that's also possible I think.

File: chapter-npm.asciidoc [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+format. You can share internally developed, proprietary packages within your organization via these private registries
+allowing you to collaborate efficiently across development teams with a central package exchange and storage location.
+
+To simplify configuration Nexus supports aggregation of npm registries. This allows you to expose all the external
+package from the npm registry and other public registries as well as the private registries as one registry, which
```
Comment (joedragons) [inline chapter-npm.asciidoc]: Should package be plaural?

File: chapter-npm.asciidoc [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+you can use packages others have created and made available in their NPM repositories by proxying them or downloading
+the packages and installing them in your own private registry for third party packages.
+
+
+[[npm-proxying-registries]]
+=== Proxying npm Registries
+
+To reduce duplicate downloads and improve download speeds for your developers and CI servers, you should proxy the
+registry hosted at -https://registry.npmjs.org[https://registry.npmjs.org]. It is accessed directly by npm
+out-of-the-box. You can also proxy any other registries you require.
```
Comment (joedragons) [inline chapter-npm.asciidoc]: MInor: I'm not sure what 'accessed directly out of the box' means.  2 cents.

File: chapter-npm.asciidoc [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+- Define 'Name'
+- Define URL for 'Remote storage'
+- Select 'Blobstore' for 'Storage'
+
+[[npm-private-registries]]
+=== Private npm Registries
+
+A private npm registry can be used to upload your own packages as well as third-party packages. You can create a private
+npm registry by setting up a hosted repository with the npm format in Nexus.It is good practice to create two separate
```
Comment (joedragons) [inline chapter-npm.asciidoc]: Line 19, npm format was +npm+ format.  Don't feel strongly but I believe is inconsistant here.
Aside: Was kinda surprised hosted didn't have markup but it is consistant.  Didn't check the writing guide...probably should reread it=)

File: chapter-npm.asciidoc [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+hosted repositories for these purposes.
+
+To create a hosted repository with npm format, simply create a new 'npm (hosted)' as documented in
```
Comment (joedragons) [inline chapter-npm.asciidoc]: Same here npm v +npm+

File: chapter-npm.asciidoc [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+configured to require authentication by adding +always-auth=true+ to the +.npmrc+ file and adding the encoded
+authentication details as documented in <<npm-deploying-packages>>.
+
+[[npm-deploying-packages]]
+=== Publishing npm Packages
+
+Publishing your own packages to a npm hosted repository allows you to share packages across your organization or with
+external partners.
+
+The npm publish command uses a +registry+ configuration value to know where to publish your package. There are several
```
Comment (joedragons) [inline chapter-npm.asciidoc]: Being an actual command should npm publish be in single quotes?  Or is that markup for Nexus stuff only?

File: chapter-npm.asciidoc [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+Once you have the encoded credentials the value as well as author information can then be added to the +.npmrc+ file:
+
+----
+init.author.name = Jane Doe
+init.author.email = jane@example.com
+init.author.url = http://blog.example.com
+# an email is required to publish npm packages
+email=jane@example.com
+_auth=YWRtaW46YWRtaW4xMjM=
+----
```
Comment (joedragons) [inline chapter-npm.asciidoc]: As noted in the testguide, I tried not having any of the init.* stuff and it works without it.  email and _auth were needed.  Feel like I might have brought this up before and these were left in, but 2 cents.

File: chapter-npm.asciidoc [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+
+TIP: Whatever tool you use to generate the encoded username and password string, try to encode the string
++admin:admin123+, which should result in +YWRtaW46YWRtaW4xMjM=+. Another example for a valid setup is
++jane:testpassword123+ resulting in +amFuZTp0ZXN0cGFzc3dvcmQxMjM=+.
+
+With this configuration you can run +npm publish+ for your package. More information about package creation can be found
+on the https://www.npmjs.org/doc/cli/npm-publish.html[npm website].
+
+Once a package is published to the private registry in Nexus, any other developers or build servers, that access Nexus
```
Comment (joedragons) [inline chapter-npm.asciidoc]: Second comma seems extraneous to me.  If you remove the text after it or the text in commas the sentence makes no sense.
aka "Once a package is published to the private registry in Nexus that access Nexus" & "Once a package is published to the private registry in Nexus, any other developers or build servers".

**Comment by joedragons** [comment]
Core looks good.  +1 (besides comments) if you intend to have other PRs to do scope / deprecate / search.  +0 otherwise.

File: chapter-npm.asciidoc (hunk 3 above)
Comment (mosabua) [inline chapter-npm.asciidoc]: Yes, fixed

File: chapter-npm.asciidoc (hunk 4 above)
Comment (mosabua) [inline chapter-npm.asciidoc]: Reworded.

File: chapter-npm.asciidoc (hunk 5 above)
Comment (mosabua) [inline chapter-npm.asciidoc]: hosted is just a normal word in this context.

File: chapter-npm.asciidoc (hunk 6 above)
Comment (mosabua) [inline chapter-npm.asciidoc]: +npm+ is used when it is a command line example. UI labels and such are referenced with 'label'.

File: chapter-npm.asciidoc (hunk 8 above)
Comment (mosabua) [inline chapter-npm.asciidoc]: I didnt know. Removed.

**Comment by mosabua** [comment]
We will aggregate it all in this branch since we can only merge this closer to when we release m6 .. we might push other things live earlier..

File: chapter-npm.asciidoc (hunk 6 above)
Comment (joedragons) [inline chapter-npm.asciidoc]: Okie, then I think line 19 is wrong but I'll leave to you since I feel just as likely I'm still confused=)

Comment (mosabua) [inline chapter-npm.asciidoc]: I suppose .. kind of an edge case there. I will change it to follow your line of thinking there.

**Comment by mosabua** [comment]
This is merged into the nexus-3-m6 manually so this PR can be closed.

**Pull request closed without merging**